File: Java/src/main/java/com/ppdai/open/core/ObjectDigitalSignHelper.java
package com.ppdai.open.core;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Created by xuzhishen on 2016/3/11.
 */
public class ObjectDigitalSignHelper {

    /**
     * 获取待签名对象的排序字符串
     * @param propertyObjects
     * @return
     */
    public static String getObjectHashString(PropertyObject ...propertyObjects){
        List<String> list = new ArrayList<String>();
        for (PropertyObject propertyObject : propertyObjects){
            if (propertyObject.isSign())
                list.add(propertyObject.getLowerName());
        }
        Collections.sort(list);

        StringBuffer sb = new StringBuffer();
        for (String ln : list){
            for (PropertyObject propertyObject : propertyObjects){
                if (ln.equals(propertyObject.getLowerName())){
                    sb.append(propertyObject.toString());
                    break;
                }
            }
        }
        return sb.toString();
    }
}

File: C#/OpenApi.Utility/ObjectDigitalSignHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;

namespace OpenApi.Utility
{
    public class ObjectDigitalSignHelper
    {
        /// <summary>
        /// 获取签名字符串排序
        /// </summary>
        /// <param name="propertyObjects">属性对象</param>
        /// <returns></returns>
        public static string GetObjectHashString(params PropertyObject[] propertyObjects)
        {
            StringBuilder sb = new StringBuilder();
            var sortPropertys = propertyObjects.Where(p => p.IsSign).OrderBy(p => p.LowerName).GetEnumerator();
            while (sortPropertys.MoveNext()) {
                sb.Append(sortPropertys.Current.ToString());
            }
            return sb.ToString();
        }
    }
}
